compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -556,18 +556,93 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(using Context) {
             ref(lastSelf).outerSelect(lastLevel - level, selfSym.info)
           else
             inlineCallPrefix
-      val binding = ValDef(selfSym.asTerm, QuoteUtils.changeOwnerOfTree(rhs, selfSym)).withSpan(selfSym.span)
+      val binding = accountForOpaques(
+        ValDef(selfSym.asTerm, QuoteUtils.changeOwnerOfTree(rhs, selfSym)).withSpan(selfSym.span))
       bindingsBuf += binding
       inlining.println(i"proxy at $level: $selfSym = ${bindingsBuf.last}")
       lastSelf = selfSym
       lastLevel = level
     }
   }
 
+  /** A list of pairs between TermRefs appearing in thisProxy bindings that
+   *  refer to objects with opaque type aliases and local proxy symbols
+   *  that contain refined versions of these TermRefs where the aliases
+   *  are exposed.
+   */
+  private val opaqueProxies = new mutable.ListBuffer[(TermRef, TermRef)]
+
+  /** Map first halfs of opaqueProxies pairs to second halfs, using =:= as equality */
+  def mapRef(ref: TermRef): Option[TermRef] =
+    opaqueProxies
+      .find((from, to) => from.symbol == ref.symbol && from =:= ref)
+      .map(_._2)
+
+  /** If `binding` contains TermRefs that refer to objects with opaque
+   *  type aliases, add proxy definitions that expose these aliases
+   *  and substitute such TermRefs with theproxies. Example from pos/opaque-inline1.scala:
+   *
+   *  object refined:
+   *    opaque type Positive = Int
+   *    inline def Positive(value: Int): Positive = f(value)
+   *    def f(x: Positive): Positive = x
+   *  def run: Unit = { val x = 9; val nine = refined.Positive(x) }
+   *
+   *  This generates the following proxies:
+   *
+   *    val $proxy1: refined.type{type Positive = Int} =
+   *      refined.$asInstanceOf$[refined.type{type Positive = Int}]
+   *    val refined$_this: ($proxy1 : refined.type{Positive = Int}) =
+   *      $proxy1
+   *
+   *  and every reference to `refined` in the inlined expression is replaced by
+   *  `refined_$this`.
+   */
+  def accountForOpaques(binding: ValDef)(using Context): ValDef =
+    binding.symbol.info.foreachPart {
+      case ref: TermRef =>
+        for cls <- ref.widen.classSymbols do
+          if cls.containsOpaques && mapRef(ref).isEmpty then
+            def openOpaqueAliases(selfType: Type): List[(Name, Type)] = selfType match
+              case RefinedType(parent, rname, TypeAlias(alias)) =>
+                val opaq = cls.info.member(rname).symbol
+                if opaq.isOpaqueAlias then
+                  (rname, alias.stripLazyRef.asSeenFrom(ref, cls))
```

Comment:
Minor point: we could check the accessibility of `alias` at this point to generate a slightly better error message if it isn't accessible, but the current one isn't too bad:
```scala
object refined:
  private class Internal(x: Int)
  opaque type Positive = Internal
  inline def Positive(value: Int): Positive = new Internal(value)

object test:
  def run: Unit =
    val x = 9
    val nine = refined.Positive(x)
```
```scala
-- Error: tests/pos/opaque-inline1.scala:10:31 ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
10 |    val nine = refined.Positive(x)
   |               ^^^^^^^^^^^^^^^^^^^
   |               class Internal cannot be accessed as a member of (refined : refined.type{Positive = refined.Internal}) from module class test$.
   | This location contains code that was inlined from opaque-inline1.scala:4
```

Reply:
There was a wrong condition for eliding the proxies, which is fixed in 2414033. Now pos/opaque-inline1 does not compile anymore if I remove the cast. It fails in -Ycheck with
```
java.lang.AssertionError: assertion failed: Found:    Int
Required: refined.Positive
found: class Int in package scala with class Int, flags = final abstract <scala-2.x> <touched>, underlying = Int, AnyVal {...}
expected: type Positive in object refined with type Positive, flags = <deferred> opaque <touched>, underlying = refined.Positive, , Any, {...}
tree = {
  val $proxy1: refined.type{Positive = Int} = 
    refined.$asInstanceOf[refined.type{Positive = Int}]
  val refined$_this: ($proxy1 : refined.type{Positive = Int}) = $proxy1
  refined$_this.f(x):refined$_this.Positive
} while compiling opaque-inline1.scala
```
The weird thing is that with the original condition everything succeeded, even though the expansion was then
```scala
  val x = 9
  val nine = refined.f(x): refined.Positive
```
This is clearly type-incorrect, since `9` is an `Int` and `refined.f` expects as `refined.Positive`. Indeed, if I write this expansion manually, it fails.
```scala
-- [E007] Type Mismatch Error: opaque-inline1.scala:12:25 ----------------------
12 |    val nine = refined.f(x): refined.Positive
   |                         ^
   |                         Found:    (x : Int)
   |                         Required: refined.Positive
```
But somehow `-Ycheck` does not test this. I traced Ychecks behavior and it seems to think that `refined.f: Int => Int` even though it should be `refined.f: Positive => Positive`. The reason for this is that it copies the tree `refined.f` as is from the inline method, which is in a scope where `Positive = Int`. Hence, `Int => Int` is a legal type for `f` there. It's the copying that is wrong, and unfortunately Ycheck does not catch it.

But the cases won't arise anymore with this PR.